unixfs/examples/resolve.rs

```diff
@@ -0,0 +1,291 @@
+            Found(cid) => {
+                // either root was a Directory or we got lucky with a HAMT directory.
+                // With HAMTDirectories the top level can contain a direct link to the target, but
+                // it's more likely it will be found under some bucket, which would be the third
+                // case in this match.
+                println!("got lucky: found {} for {:?}", cid, segment);
+                println!("cache {:?}", cache);
+                root = cid;
+                continue;
+            }
+
+            NotFound => return Ok(None),
+
+            // when we stumble upon a HAMT shard, we'll need to look up other blocks in order to
+            // find the final link. The current implementation cannot lookup the directory by
+            // hashing the name and looking it up, but the implementation can be changed underneath
+            // but the API would stay the same.
+            //
+            // HAMTDirecotories or HAMT shards are multi-block directories where the entires are
+            // bucketed per their hash value.
+            NeedToLoadMore(walker) => walker,
+        };
+
+        eprintln!("walking {} on {:?}", root, segment);
+
+        let mut other_blocks = 1;
+
+        loop {
+            let (first, _) = walker.pending_links();
+            buf.clear();
+            eprintln!("  -> reading {} while searching for {:?}", first, segment);
+            blocks.as_file(&first.to_bytes())?.read_to_end(&mut buf)?;
+
+            match walker.continue_walk(&buf, &mut cache)? {
+                NotFound => {
+                    println!("cache {:?}", cache);
+                    return Ok(None);
+                }
+                Found(cid) => {
+                    eprintln!(
+                        "     resolved {} from {} after {} blocks to {}",
+                        segment, root, other_blocks, cid
+                    );
+                    root = cid;
+                    break;
+                }
+                NeedToLoadMore(next) => walker = next,
+            }
+            other_blocks += 1;
+        }
+    }
+
+    println!("cache {:?}", cache);
+    Ok(Some(root))
+}
+
+enum Error {
+    OpeningFailed(IoError),
+    Other(IoError),
+    Traversal(ResolveError),
+}
+
+impl From<IoError> for Error {
+    fn from(e: IoError) -> Error {
+        Error::Other(e)
+    }
+}
+
+impl From<ResolveError> for Error {
+    fn from(e: ResolveError) -> Error {
+        Error::Traversal(e)
+    }
+}
+
+impl From<LookupError> for Error {
+    fn from(e: LookupError) -> Error {
+        Error::Traversal(e.into())
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Error::*;
+        match self {
+            OpeningFailed(e) => write!(fmt, "File opening failed: {}", e),
+            Other(e) => write!(fmt, "Other file related io error: {}", e),
```

Comment:
I see similar errors as in previous examples; perhaps in a follow-up PR we could put some of the common bits into an `examples/common.rs` module or something of this sort?

Reply:
I don't know how to do that? Can examples share code?

Reply:
As long as there is a common module; as far as I can see, there is currently a seemingly useless `main.rs` that could perhaps be turned into a common `lib.rs` and link all the example modules together; haven't played with examples in a while, but the structure shouldn't be different than more advanced tests or benches.

Reply:
I tried to keep these `unixfs/examples/` self contained which is also why they have the repeated content. Would perhaps not prefer to move them under `examples/examples/`.